### PR TITLE
Remove `yamlgen`. Refactor deployment and yaml generation to use Helm templating

### LINKED ISSUE #126: 0.2.0: Create Brupop Helm Chart
**Issue or Feature Request:**
Brupop should be installable via Helm. This would provide customers an easy route to install dependencies, like [cert-manager](https://cert-manager.io/) for managing SSL certificates used for communication with the apiserver, or the OpenTelemetry Collector, should we integrate with that rather than direct Prometheus integration.

### PR DESCRIPTION
**Pull request by jpmcb**
**Issue number:**

Closes #126 

**Description of changes:**

Refactors deployment and yaml templating to use Helm. Still a work in progress. Using this draft PR as a staging area for early feedback

**Testing done:**

Able to:
- `make` images
- `helm install brupop-crd deploy/charts/bottlerocket-shadow`
- `helm install brupop deploy/charts/bottlerocket-update-operator`

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

**Comment by jpmcb** [comment]
Force pushed to update certificate resources that changed with #340 - Ran through successful updates using:

```
$ helm install brupop-crd --create-namespace deploy/charts/bottlerocket-shadow
$ helm install brupop deploy/charts/bottlerocket-update-operator
```

**Comment by jpmcb** [comment]
Force pushed:
- fixed build errors
- resolved merge conflict (with `Cargo.lock` update)
- Updated READMEs with documentation on how to install brupop via helm
- added `make manifest` target that can be used to generate manifest at root of repository
- Removed `.env` file (in favor of helm values yaml file)
- Added logic to create / delete brupop namespace in integration tests

Marking this as ready for review.

There are no big things left to do here and I'd love to start getting some feedback from the community. I'm sure there are little things left here since this is a bigger refactor, but we can continue to tackle those in the coming weeks!

We'll also want to think about doing a one off change to the new release workflow to ensure that works for releases. Aka, we can change that to trigger on each push to `develop` and ensure it's working properly. That workflow is designed to push a helm repo to a github page hosted in this repository. We'll also need to ensure we have the right access right associated with that so it can be created on release automatically.

File: deploy/charts/bottlerocket-shadow/templates/NOTES.txt
Comment (stmcginnis) [inline deploy/charts/bottlerocket-shadow/templates/NOTES.txt]: s/affectionatly/affectionately/

File: README.md
Comment (stmcginnis) [inline README.md]: nit: "it's" means "it is", "its" is possessive.

**Review by stmcginnis — COMMENTED** [review]
Looks good to me. There are a lot of changes here, but with the noted testing done I feel comfortable with the change.

Looks like there is a merge conflict to be resolved now. Since an update is needed anyway, a couple small things pointed out.

File: README.md
Comment (cbgbt) [inline README.md]: I might clean up some of the language on these comments since they are lifted from what was previously a dev-level config file, but e.g. here we should probably clarify that this uses `"node.kubernetes.io/exclude-from-external-load-balancers"` labels under the hood.

File: deploy/charts/bottlerocket-update-operator/templates/NOTES.txt
Comment (cbgbt) [inline deploy/charts/bottlerocket-update-operator/templates/NOTES.txt]: nit: clean up wording to clarify which API is being discussed.

```suggestion
The bottlerocket-update-operator will now begin querying Bottlerocket's update API for updates.
```

File: README.md
Comment (cbgbt) [inline README.md]: Why separate the CRD and the operator installations?

File: deploy/build.rs
Comment (cbgbt) [inline deploy/build.rs]: Since we're still using dotenv as a development tool, should we leave the `.env` file around with some comments for what it can do?

File: deploy/charts/bottlerocket-shadow/README.md
Comment (cbgbt) [inline deploy/charts/bottlerocket-shadow/README.md:8]: May want to additionally leave a note here about labeling hosts so that they receive a Shadow.

File: deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml
Comment (cbgbt) [inline deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml:1]: Possibly I'll answer this myself as I keep reading, but in the pre-helm setup, this was autogenerated based on our Rust models. Do we manually keep the two in sync now, or is it still possible to generate this document?

File: deploy/charts/bottlerocket-update-operator/README.md
Comment (cbgbt) [inline deploy/charts/bottlerocket-update-operator/README.md:6]: Is it possible to model this dependency using Helm itself?

File: integ/src/main.rs
Comment (cbgbt) [inline integ/src/main.rs]: Should we exit here? I could see this being non-fatal if you get this error when such a namespace already exists, but otherwise it seems like we should return an error here.

File: deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml
Comment (cbgbt) [reply deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml:1]: Just to flesh this feedback out, a lot of these values only make sense in the context of the Kubernetes YAML model and have no bearing on the `brupop` sources, seems totally fine to remove the Rust models.

What I'm mostly interested in is the auto-generated enums and validations based on our Rust models, which was pretty handy. It would be a lot easier to have a mismatch or mistake if we untie them. Reading more carefully, it seems like the shadow CRD is still autogenerated (probably for this reason!). Nice!

**Comment by cbgbt** [comment]
This is awesome, nice work @jpmcb! 🚀

**Review by cbgbt — COMMENTED** [review]
Whoops, forgot to hit send on these.

File: README.md
Comment (jpmcb) [reply README.md]: Good question!

This is a common Helm convention: the CRD chart is separate from the consumer of the CRD. This is because helm will reconcile resources from a operator chart in the most efficient way possible:

> Helm is optimized to load as many resources into Kubernetes as fast as possible. By design, Kubernetes can take an entire set of manifests and bring them all online (this is called the reconciliation loop).

> But there's a difference with CRDs.

> For a CRD, the declaration must be registered before any resources of that CRDs kind(s) can be used. And the registration process sometimes takes a few seconds.

[Helm Docs - Install a CRD Declaration Before Using the Resource](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#install-a-crd-declaration-before-using-the-resource)

There _is an option_ in Helm to let the `helm install` handle this for you. However, this has the follow caveats:

> There is no support at this time for upgrading or deleting CRDs using Helm. This was an explicit decision after much community discussion due to the danger for unintentional data loss. Furthermore, there is currently no community consensus around how to handle CRDs and their lifecycle. As this evolves, Helm will add support for those use cases.

[Helm docs - Some caveats (and explanations)](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#some-caveats-and-explanations)

Instead, I seperated the two charts to make handling the bottlerocket shadows explicit:
- Cluster operators can can manage the `brs` CRD out of band from the operator itself
  - We avoid the problem of not being able to upgrade/downgrade CRD versions (even though brs v2 is relatively stable)

There is also an argument that has been made in the community that cluster custom resource definitions should only be installable by a cluster admin (where strong RBAC rules are often enforced) but other users may have access to view and edit the `brs` resources and namespace.

In short, this was a deliberate decision to enforce stronger RBAC practices for our users and gets around some issues with versioning CRDs in helm.

File: deploy/build.rs
Comment (jpmcb) [reply deploy/build.rs]: Hmmmm that's a good point - I think this means we'll have to yes. Let me circle back and look at that.

File: deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml
Comment (jpmcb) [reply deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml:1]: > What I'm mostly interested in is the auto-generated enums and validations based on our Rust models, which was pretty handy. It would be a lot easier to have a mismatch or mistake if we untie them. Reading more carefully, it seems like the shadow CRD is still autogenerated (probably for this reason!). Nice!

Right - I couldn't 100% decouple the kubernetes resources from the rust code. However, I think I struck a balance between what our rust code needs in it (like the enums, etc.) and yaml kubernetes resources

File: deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml
Comment (jpmcb) [inline deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml]: This needs to be templated with the values defined in the `values.yaml` for this chart 👀

File: deploy/charts/bottlerocket-update-operator/README.md
Comment (jpmcb) [reply deploy/charts/bottlerocket-update-operator/README.md:6]: > Is it possible to model this dependency using Helm itself?

Good question.

I think [this doc from the helm maintainers on the difficulty handling CRDs ](https://github.com/helm/community/blob/f9e06c16d89ccea1bea77c01a6a96ae3b309f823/architecture/crds.md)is really good prior-art to this PR.

Generally, I believe our posture should be to make few assumptions about the shape of our user's clusters and how they are administered. 

I'll annotate [the following section from the doc](https://github.com/helm/community/blob/f9e06c16d89ccea1bea77c01a6a96ae3b309f823/architecture/crds.md#installing-crds) to give some insight into what I believe makes sense for brupop:

> There are a number of well-described issues with installing CRDs. Users must have broad permissions. A CRD must be fully installed before a manifest that references that CRD can be accepted by the Kubernetes API server. (However, a CR can be accepted before there is anything available to handle the CR event.) It is entirely possible for two different applications to require different versions of the same CRD, but no clear way to support or declare that need within Kubernetes. This is exacerbated by the fact that while a CRD is a global object, the controllers that listen for CRD events may be namespace-bound. This means that two instances of the same controller (in different namespaces) can use the same CRD definition. There is no way to query Kubernetes to discover this fact.

Brupop _does_ have namespaced resources (its API server, controller, daemonsets, etc.) that rely on the global `brs` CRD to be present before install. We shouldn't make an assumption on the elevated permission of the users deploying or accessing the brupop namespaced resources (for example, creating namespaces may be a permission many cluster users don't have).

And while I'd be surprised if anyone was attempting to use `brs` as a _dependent_ resource in their own custom kubernetes controller, it is ultimately possible. And I believe the really key part of this is "There is no way to query Kubernetes to discover this fact." We can't make assumptions _from the brupop perspective or install_ on if some other namespaced resources are using the `brs` custom resources. We don't want to couple the install of the CRD and the controller too tightly since we do provide multiple versions of the CRD. Therefore, the CRD should be it's own, separate chart.

> All of these things present challenges for installing CRDs. Essentially, any installation process must have some kind of logic like "If the CRD exists, check to make sure that it is compatible with the version of the CRD that I need, and then proceed on my installation. Otherwise, attempt to install the CRD, and handle the case that I might not have permission to do so."

> After looking at several ways of making it easy for chart developers to do this, we alighted on a fairly easy (though inelegant) solution: Helm, as it stands today, provides a separate location in a chart to add CRDs. These will be uploaded before the chart's templates are even rendered (thus avoiding verification issues with the discovery API). If the CRD exists or if the permissions do not allow it, the chart will fail before the application is deployed.

> Of course, users are unhappy with this for a host of reasons. They want CRDs templated (without understanding the race conditions). They want stronger version controls. They don't like having a separate directory for CRDs. We sympathize... but we currently have not devised a better solution for the aforementioned problems.

And unfortunately, this is exactly where we find ourselves; CRD within a separate chart. Our CRD _does_ need to be templated (in order to derive the namespace and port of the conversion webhook) and we can't rely on pre-rendered templating from the controller chart.

Now to your question:

---

> model this dependency using Helm itself?

Modeling this dependency in Helm would require we make the CRD apart of the update operator's chart itself. This is the last point in the above deposition `Helm provides a separate location in a chart to add CRDs` where a CRD is installed before helm renders and installs the remaining resources.

But, as laid out, I don't believe coupling the CRD and the controller install is the way we want to go.

The alternative is a fast fail when attempting to `helm install` the brupop controller indicating the CRD is not present on the cluster.

**Comment by jpmcb** [comment]
Force pushed to correct for `Cargo.lock` merge conflict.

Going to do another round of iteration here based on comments and a few `TODO` lines left.

File: integ/src/main.rs
Comment (jpmcb) [reply integ/src/main.rs]: > I could see this being non-fatal if you get this error when such a namespace already exists

Exactly. I ran into a few cases where the integration tests didn't completely remove the namespace (or there were hanging resources preventing the namespace from being purged, my temporary cluster creds timed-out, etc. etc.) and it was just easier to warn instead of completely bail out of creation.

But if we think this is too loose, I can also see exiting early here and requiring the operator to fix the problem.

**Comment by jpmcb** [comment]
Force pushed to add the following:
- Snapshot testing of our generated yaml CRDs. This way, if something changes under the hood in our rust CRD definitions, we can have a signal to accept/reject those changes. This mechanism can also be used as an indicator to update the CRD templates in the helm chart.
  - `cargo test` to run snapshot tests via `insta`. And `cargo insta review` to accept new snapshots.
  - `make check-crd-golden-diff` to compare the rust generated CRD yaml to the helm chart template.
- Remove `const` declaration of namespace. With the helm chart, removes the requirement that users deploy into the `bottlerocket-aws` namespace we had hard coded. Now, we infer the current _in cluster_ namespace from configs.
- Update docs based on feedback

---

Integration tests still looking 👍🏼 

```
❯ k get bottlerocketshadows -A
NAMESPACE                 NAME                                                STATE   VERSION   TARGET STATE   TARGET VERSION   CRASH COUNT
brupop-bottlerocket-aws   brs-ip-192-168-142-119.us-west-2.compute.internal   Idle    1.11.1    Idle                            0
brupop-bottlerocket-aws   brs-ip-192-168-144-207.us-west-2.compute.internal   Idle    1.11.1    Idle                            0
brupop-bottlerocket-aws   brs-ip-192-168-156-62.us-west-2.compute.internal    Idle    1.11.1    Idle                            0
```

File: integ/src/main.rs
Comment (cbgbt) [reply integ/src/main.rs]: It might be worth leaving a comment or somehow otherwise making it clear in the code that this is OK in the case that the namespace exists.

File: deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml
Comment (jpmcb) [reply deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml:1]: Yes - let's move these togeher

**Comment by jpmcb** [comment]
Force pushed to template a few remaining references to `brupop-bottlerocket-aws` - getting close!!

File: .github/workflows/deploy-helm-charts.yaml
Comment (jpmcb) [inline .github/workflows/deploy-helm-charts.yaml:9]: Maybe for this PR, we can change this [to include `workflow_dispatch`](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#workflow_dispatch). That way, the maintainers with access rights can test this workflow with `gh` on the command line without having to cut arbitrary releases:

```
gh workflow run deploy-helm-chart.yml
```

Comment (cbgbt) [reply .github/workflows/deploy-helm-charts.yaml:9]: I am okay with including this in a separate PR, we just want it before we cut this into a release.

File: README.md
Comment (cbgbt) [reply README.md]: This makes sense, thanks for elaborating.

File: agent/src/agentclient.rs
Comment (cbgbt) [inline agent/src/agentclient.rs]: nit: Does it make sense to have a `infer_namespace()` function Rather than repeating this?

File: apiserver/src/client/webclient.rs
Comment (cbgbt) [inline apiserver/src/client/webclient.rs]: These should be `&str` since this function does not need to take ownership of the string/force re-allocation.

Note: You can also do `T: AsRef<str>` if you want to accept both `&str` and `String`, but many APIs just take `&str` unless they need ownership.

File: apiserver/src/client/webclient.rs
Comment (cbgbt) [inline apiserver/src/client/webclient.rs]: Rather than add the namespace to the trait, I think we should just take it as an argument on the constructor for `K8SAPIServerClient` and store it -- it's not generically necessary for all implementations.

File: controller/src/main.rs
Comment (cbgbt) [inline controller/src/main.rs:44]: Maybe throwing this under something like `models/util.rs` or `models/k8s_utils.rs` could help us code share.

File: deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml
Comment (cbgbt) [reply deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml:1]: I could be missing where this is already done while writing this comment, so apologies if so, but while it's on my mind: I like the `insta` approach to managing these files. Would be good to document that, possibly in the Development section of the README.

File: deploy/charts/bottlerocket-update-operator/README.md
Comment (cbgbt) [reply deploy/charts/bottlerocket-update-operator/README.md:6]: Thanks for digging in here. The relationships are a bit awkward, but it does seem like the best setup we can muster using Helm.

File: deploy/src/test.rs
Comment (cbgbt) [inline deploy/src/test.rs]: Ah, here's the docs I was asking about. Nice. You'll see this when the test fails and know what to do next.

Comment (cbgbt) [reply deploy/src/test.rs]: nit: You should be able to move this under `deploy/tests/generated_crds.rs` or something similar. See the Rust docs about [integration tests](https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html).

**Review by cbgbt — COMMENTED** [review]
Looks good! Just a few minor changes that I'd like to see and a few nits that I could take or leave.

File: agent/src/agentclient.rs
Comment (jpmcb) [reply agent/src/agentclient.rs]: We do this elsewhere where we explicitly create an incluster DNS client config:
https://github.com/bottlerocket-os/bottlerocket-update-operator/blob/f69a1708c3fb095b160665ca162fe4d2f602d6bc/apiserver/src/main.rs#L48-L49

And this is a workaround until we can get the next version of `kube-rs` with the `default_namespace()` function:
https://github.com/kube-rs/kube/pull/1123

**Comment by jpmcb** [comment]
Force pushed to address merge conflicts - addressing comments next!

File: apiserver/src/client/webclient.rs
Comment (jpmcb) [reply apiserver/src/client/webclient.rs]: > it's not generically necessary for all implementations

Won't all implementations of `update_bottlerocket_shadow` require the namespace since the `bottlerocketshadow` is a namespaced CRD?

We got away with this before because the controller _only_ knew how to operate on `brs` in the hard coded `brupop-bottlerocket-aws` namespace. If a cluster operator tried to deploy `brs` resources to any other namespace, the entire stack breaks.

In my mind, because the CRD is namespaced, `brs` resources are tightly coupled to the namespace they are deployed in and the API should reflect that as well

File: agent/src/agentclient.rs
Comment (jpmcb) [reply agent/src/agentclient.rs]: Thinking alittle more on this: we'll still need to get an incluster config so I wonder if a subsequent PR (once `default_namespace()` is available in `kube-rs`) would be to refactor these calls into a common shared library

File: deploy/charts/bottlerocket-shadow/README.md
Comment (jpmcb) [reply deploy/charts/bottlerocket-shadow/README.md:8]: Done!!

File: deploy/build.rs
Comment (jpmcb) [reply deploy/build.rs]: The new workflow should become:

1. Create a `values.yaml` file (maybe we can create a top level one for development purposes and git ignore it)
2. Populate it with necessary values. For example:
```yaml
namespace: "my-custom-ns"
```
3. Pass the values yaml file to be used during install either via:
- `helm install -f values.yaml ...` (to install to a cluster)
- `helm template -f values.yaml ... > file.yaml` (to template into the yaml)

I think there's some better dev UX to tease out here, especially with how the integration tests work into this flow: currently, the integration tests will just read from `bottlerocket-update-operator.yaml` and apply that to the cluster but there's opportunity here to make this more seamless.

File: integ/src/main.rs
Comment (jpmcb) [reply integ/src/main.rs]: Circling back on this - ended up removing `process_brupop_namespace` since the final generated yaml manifest has the namespace resource 👍🏼

File: deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml
Comment (jpmcb) [reply deploy/charts/bottlerocket-shadow/templates/custom-resource-definition.yaml]: Done! ☑️

File: apiserver/src/client/webclient.rs
Comment (jpmcb) [reply apiserver/src/client/webclient.rs]: Done!!! ✅

File: deploy/src/test.rs
Comment (jpmcb) [reply deploy/src/test.rs]: Done!! ✅

File: apiserver/src/client/webclient.rs
Comment (cbgbt) [reply apiserver/src/client/webclient.rs]: The way I'm thinking about it is that the namespace here identifies the location of the APIServer itself, so it's more of a client configuration detail. It's contrived for sure, but the test implementation is an example of one that doesn't need the configuration, but would now be required to ignore `namespace` or assert that the value corresponds to an environment-derived value.

File: agent/src/agentclient.rs
Comment (cbgbt) [reply agent/src/agentclient.rs]: I would prefer if we clean this up now given how many times we repeat this pattern. It would make the move to `default_namespace()` a smaller change as well.

**Comment by jpmcb** [comment]
Force pushed to address above comments

File: apiserver/src/client/webclient.rs
Comment (cbgbt) [reply apiserver/src/client/webclient.rs]: > In my mind, because the CRD is namespaced, brs resources are tightly coupled to the namespace they are deployed in and the API should reflect that as well

Isn't this also reflected if we make it a member of the client constructor? It's not possible (per our rbac permissions) for the API server to modify shadows outside of its namespace, I don't think there's ever a case where you pass a different namespace in this value that succeeds.

File: agent/src/agentclient.rs
Comment (jpmcb) [reply agent/src/agentclient.rs]: Sounds good!! 😄

File: apiserver/src/client/webclient.rs
Comment (jpmcb) [reply apiserver/src/client/webclient.rs]: >  the test implementation is an example of one that doesn't need the configuration, but would now be required to ignore namespace or assert that the value corresponds to an environment-derived value.

That's an excellent point!!! Yes, this can be refactored out to the constructor!

Comment (jpmcb) [reply apiserver/src/client/webclient.rs]: Ended up doing abit of a refactor. Now, when the various client configs are created, we pass down the namespace into subsequent constructors that require namespaced resources. It didn't make sense to have to create a new client config with each query to namespaced resources. That would be a costly operation. Instead, we can now access the `namespace` data member where it makes sense.

**Comment by jpmcb** [comment]
Force push to address Sean's comments around infering the namespace throughout. Now, when the various client configs are created, we pass down the namespace into subsequent constructors that require namespaced resources. It didn't make sense to have to create a new client config with each query to namespaced resources. That would be a frequent, costly operation. Instead, we can now access the namespace data member built from `new`.

**Review by cbgbt — APPROVED** [review]
Super clean! Nice work John! 🚀 🎉

File: Makefile
Comment (webern) [inline Makefile:98]: Does this require a kubeconfig or Kubernetes cluster? If not, this is cool...

File: integ/src/main.rs
Comment (webern) [inline integ/src/main.rs:222]: If the user can configure the namespace of their choice, how is it that you get that from the kubeconfig?

File: models/src/constants.rs
Comment (webern) [inline models/src/constants.rs]: Is this TODO intentional?

File: deploy/tests/snapshots/generated_crds__generated_crds.snap
Comment (webern) [inline deploy/tests/snapshots/generated_crds__generated_crds.snap]: Is this filename as intended? `generated_crds__generated_crds.snap`

**Review by webern — APPROVED** [review]
About 1/3 coverage from me. A few questions of ignorance on my part. Looks good.

File: Makefile
Comment (jpmcb) [reply Makefile:98]: Right, no kubeconfig/kubernetes cluster needed. It just generates the yaml for us. For example, here, I unset my context (so I'm not targeting any cluster) and I can still generate the yaml:

```
❯ k config unset current-context
Property "current-context" unset.

❯ kubectl create namespace test-namespace --dry-run=client -o yaml
apiVersion: v1
kind: Namespace
metadata:
  creationTimestamp: null
  name: test-namespace
spec: {}
status: {}
```

File: integ/src/main.rs
Comment (jpmcb) [reply integ/src/main.rs:222]: Good question: the `config.default_namespace` property is the namespace defined by the loaded client config. This is typically the namespace of where the execution is currently happening.

So, in the case of the controller, it reads the in-cluster config and infers that it is in the "bottlerocket-aws" (or whatever the namespace was set during deploy).

This case in the integration tests is abit different in that we aren't inferring a in-cluster client config, we load one explicitly from eksctl. But the principle is the same - it infers the default namespace to deploy some stuff to (and in this case, it is actually "default")

It's abit confusing because the name "default" is overloaded. Most of the time, from the controllers perspective, we are simply attempting to infer what namespace the current context is executing from.

File: models/src/constants.rs
Comment (jpmcb) [reply models/src/constants.rs]: Probably not - I need to go back again after the cron PR merges to do another round of refactoring / rebase to get those into the helm chart

File: deploy/tests/snapshots/generated_crds__generated_crds.snap
Comment (jpmcb) [reply deploy/tests/snapshots/generated_crds__generated_crds.snap]: Yes. `insta` first uses of the name of the Rust file where the test originated from and the name of the test itself. I'll change the name of that file so it's alittle more clear

Comment (jpmcb) [reply deploy/tests/snapshots/generated_crds__generated_crds.snap]: The name of this file should now be `insta_test__generated_crds.snap` which is the `<test-filename>__<test-name>.snap`

**Comment by jpmcb** [comment]
Force pushed to rename the `insta` Rust test file in order to be abit more clear what's happening. The name of this file should now be `insta_test__generated_crds.snap` which is the `<test-filename>__<test-name>.snap`

FYI - this will also need a hefty rebase and abit of additonal legwork to get in once the cron PR merges

**Comment by nalshamaajc** [comment]
I hope this helm chart gets ready soon! 🙏 😄

**Comment by lsobowalejc** [comment]
Yes! this is definitely a needed feature. Can't wait!

**Comment by maxres-ch** [comment]
Hi there. Just wanted to check if there's any insight on when this might be merged?

**Comment by jpmcb** [comment]
Force pushed with large rebase (which included integrating the work for the scheduler cron) and some docs updates. The biggest thing you'll notice is different is the templates now include the `scheduler_cron_expression` template value.

```diff
diff --git a/bottlerocket-update-operator.yaml b/bottlerocket-update-operator.yaml
index cf982ec..a96c65a 100644
--- a/bottlerocket-update-operator.yaml
+++ b/bottlerocket-update-operator.yaml
@@ -711,10 +711,8 @@ spec:
                   fieldPath: spec.nodeName
             - name: MAX_CONCURRENT_UPDATE
               value: "1"
-            - name: UPDATE_WINDOW_START
-              value: "0:0:0"
-            - name: UPDATE_WINDOW_STOP
-              value: "0:0:0"
+            - name: SCHEDULER_CRON_EXPRESSION
+              value: "* * * * * * *"
           image: public.ecr.aws/bottlerocket/bottlerocket-update-operator:v1.0.0
           name: brupop
       priorityClassName: brupop-controller-high-priority

```

### Testing

1. Deployed new nodegroup to a k8s cluster

2. Installed shadow CRD successfully:
```
$ helm install brupop-crd deploy/charts/bottlerocket-shadow
...

$ kubectl get crds
NAME                                          CREATED AT
bottlerocketshadows.brupop.bottlerocket.aws   2023-06-01T18:24:49Z
...
```

3. Installed the controller stack:
```
$ kubectl create namespace brupop-bottlerocket-aws
$ helm install brupop deploy/charts/bottlerocket-update-operator
...
```

4. Nodes update as expected.

This should be ready to go in now

**Comment by gthao313** [comment]
Follow the instruction above, it works perfectly on my side! nice work!!!

**Review by gthao313 — APPROVED** [review]

File: .github/workflows/deploy-helm-charts.yaml
Comment (stmcginnis) [inline .github/workflows/deploy-helm-charts.yaml:33]: Is the plan to enable GitHub Pages to host the published charts?

File: deploy/charts/bottlerocket-update-operator/README.md
Comment (stmcginnis) [inline deploy/charts/bottlerocket-update-operator/README.md]: not: s/helmchart/helm chart/

File: .github/workflows/deploy-helm-charts.yaml
Comment (jpmcb) [reply .github/workflows/deploy-helm-charts.yaml:33]: Exactly!

Comment (jpmcb) [reply .github/workflows/deploy-helm-charts.yaml:33]: It'll require some iteration to get it right before running the next release:
- I'm not sure if pages for this repo are turned on. But we can use the `workflow-dispatch` method to launch this workflow out of band of a release.
- We'll need to update release processes to capture updating the helm chart metadata (version etc.) alongside other things (like updating the readme).

But really, in theory, it should _just work_.

**Review by stmcginnis — APPROVED** [review]
Overall looks good, and I think it would be great to get this in and available to get more usages/feedback.

**Review by stmcginnis — APPROVED** [review]
Thanks! Looks like the checks might just need a re-run to get by a transient error.

**Comment by jpmcb** [comment]
Force pushed to rebase on `develop` which now has fixes for problems caused by #461

**Review by stmcginnis — APPROVED** [review]
Let's gooooo!! 🥳

**Pull request merged**
